diff --git a/libavformat/flvenc.c b/libavformat/flvenc.c
index 4e79b06499..13e06aedbb 100644
--- a/libavformat/flvenc.c
+++ b/libavformat/flvenc.c
@@ -1067,11 +1067,6 @@ static int flv_write_packet(AVFormatContext *s, AVPacket *pkt)
                             || par->codec_id == AV_CODEC_ID_AC3
                             || par->codec_id == AV_CODEC_ID_EAC3;
 
-    if (par->codec_type == AVMEDIA_TYPE_AUDIO && !pkt->size) {
-        av_log(s, AV_LOG_WARNING, "Empty audio Packet\n");
-        return AVERROR(EINVAL);
-    }
-
     if (extended_audio)
         flags_size = 5;
     else if (par->codec_id == AV_CODEC_ID_VP6F || par->codec_id == AV_CODEC_ID_VP6A ||
@@ -1146,8 +1141,6 @@ static int flv_write_packet(AVFormatContext *s, AVPacket *pkt)
     case AVMEDIA_TYPE_AUDIO:
         flags = get_audio_flags(s, par);
 
-        av_assert0(size);
-
         avio_w8(pb, FLV_TAG_TYPE_AUDIO);
         break;
     case AVMEDIA_TYPE_SUBTITLE:
